Added map and compact filters

ilyapuchka opened the pull request.

AliSoftware commented:
Isn't that a duplicate from the `map` we have in StencilSwiftKit? I'm all for bringing our generic filters like this one in Stencil proper, but be sure they won't interfere and maybe keep the syntax we used if it's indeed the same idea?

_(Disclaimer: I'm on vacation and only skimming thru the issues from my phone so didn't read this in details)_

ilyapuchka commented:
@AliSoftware this is filter, you have node in StencilSwiftKit as far as I can see.

ilyapuchka commented:
@kylef I've solved conflicts so this is ready for review. I also added `filter` filter that allows using filters and boolean expressions, i.e. `items|filter:"$0|isPositive"` or `items|filter:"$0 > 0"`, and refactored a bit parsing of expressions so that it matches API for parsing filters

ilyapuchka commented:
@kylef I've made some changes here converting `filter` to `filterEach` and changing behaviour of `filter`. Now it will resolve it's argument as filter expression which will allow to apply dynamic filters. The use case for that can be using Sourcery to generated code for `Codable` and use different key coding strategies applied for different data models. It will allow to specify strategy as a filter in type annotation, i.e. `uppercaseFirstLetter` or `camelCaseToSnakeCase`. Currently in template I have to check annotations against hardcoded set of filters to be able to apply them.
If you want I can break it out in a separate PR as it is not related to other array filters added here.

ilyapuchka commented:
Extracted `filter` functionality to #203 
This now only contains filters for mapping and filtering arrays.
Can we merge this @kylef ?

IvanUshakov commented:
Do you have any news on this?

ilyapuchka commented on Sources/Extension.swift:
```diff
@@ -59,28 +63,30 @@ class DefaultExtension: Extension {
     registerFilter("indent", filter: indentFilter)
+    registerFilter("map", filter: mapFilter)
+    registerFilter("compact", filter: compactFilter)
+    registerFilter("filterEach", filter: filterEachFilter)
```
done

ilyapuchka commented on CHANGELOG.md:
```diff
@@ -12,6 +12,8 @@
   [Ilya Puchka](https://github.com/yonaskolb)
   [#178](https://github.com/stencilproject/Stencil/pull/178)
 
+- Added `map`, `compact` and `filterEach` filters
```
done

ilyapuchka commented on Sources/Extension.swift:
```diff
@@ -26,6 +26,10 @@ open class Extension {
+    filters[name] = .arguments({ value, args, _ in try filter(value, args) })
+  }
+
+  public func registerFilter(_ name: String, filter: @escaping (Any?, [Any?], Context) throws -> Any?) {
```
#203 already merged, had to resolve conflicts here

opfeffer commented:
Curious about the status on this @ilyapuchka. Would love to see this make it in! Thanks

namolnad commented:
Hi all, wanted to bump this to see if there was any chance of merging in or updating this PR to get it back to a mergeable state. Happy to help — thanks!
